src/nvim/lua/treesitter.c

```diff
@@ -377,6 +385,101 @@ static int parser_edit(lua_State *L)
   return 0;
 }
 
+static int parser_set_ranges(lua_State *L)
+{
+  if (lua_gettop(L) < 3) {
+    lua_pushstring(L, "not enough args to parser:set_ranges()");
+    return lua_error(L);
+  }
+
+  TSLua_parser *p = parser_check(L);
+  if (!p || !p->tree) {
+    return 0;
+  }
+
+  int bufnr = lua_tointeger(L, 2);
+
+  if (!lua_istable(L, 3)) {
+    lua_pushstring(L, "argument for parser:set_included_ranges() should be a table.");
+    return lua_error(L);
+  }
+
+  size_t tbl_len = lua_objlen(L, 3);
+  TSRange *ranges = xmalloc(sizeof(TSRange) * tbl_len);
+
+
+  // [ parser, ranges ]
+  for (size_t index = 0; index < tbl_len; index++) {
+    lua_rawgeti(L, 3, index + 1);  // [ parser, ranges, range ]
+
+    if (!lua_istable(L, -1)) {
+      xfree(ranges);
+      lua_pushstring(
+          L,
+          "argument for parser:set_included_ranges() should be a table of tables.");
+      return lua_error(L);
+    }
+
+    if (lua_objlen(L, -1) < 4) {
+      xfree(ranges);
+      lua_pushstring(
+          L,
+          "argument for parser:set_included_ranges() should be a table of ranges of 4 elements.");
+      return lua_error(L);
+    }
+
+    lua_rawgeti(L, -1, 1);  // [ parser, ranges, range, num ]
+    unsigned int start_row = lua_tointeger(L, -1);
+    lua_pop(L, 1);  // [ parser, ranges, range ]
+
+
+    lua_rawgeti(L, -1, 2);  // [ parser, ranges, range, num ]
+    unsigned int start_col = lua_tointeger(L, -1);
+    lua_pop(L, 1);  // [ parser, ranges, range ]
+
+    lua_rawgeti(L, -1, 3);  // [ parser, ranges, range, num ]
+    unsigned int stop_row = lua_tointeger(L, -1);
+    lua_pop(L, 1);  // [ parser, ranges, range ]
+
+    lua_rawgeti(L, -1, 4);  // [ parser, ranges, range, num ]
+    unsigned int stop_col = lua_tointeger(L, -1);
+    lua_pop(L, 1);  // [ parser, ranges, range ]
+
+    buf_T * buf = buflist_findnr(bufnr);
+
+    if (!buf) {
+      buf = curbuf;
+    }
+
+    // TODO(vigoux): For sure that's wrong, try to find a way to get the
```

Comment:
Is this TODO still valid?

Reply:
Depends if we keep calculating the byte offsets here or if we move it somewhere else.

Reply:
I will me changing things to use nodes instead of raw ranges, this will be far more flexible with nodes